add cumulative metrics granularity info (#5688)
dbt-labs/docs.getdbt.com: website/docs/docs/build/cumulative-metrics.md

:::

## Parameters

| Parameter | <div style={{width:'350px'}}>Description</div> | Type |
| --------- | ----------- | ---- |
| `name` | The name of the metric. | Required |
| `description` | The description of the metric. | Optional |
| `type` | The type of the metric (cumulative, derived, ratio, or simple). | Required |
| `label` | Required string that defines the display value in downstream tools. Accepts plain text, spaces, and quotes (such as `orders_total` or `"orders_total"`). | Required |
| `type_params` | The type parameters of the metric. Supports nested parameters indicated by the double colon, such as `type_params::measure`. | Required |
| `type_params::cumulative_type_params` | Allows you to add a `window`, `period_agg`, and `grain_to_date` configuration. Nested under `type_params`. | Optional |
| `cumulative_type_params::window` | The accumulation window, such as 1 month, 7 days, 1 year. This can't be used with `grain_to_date`. | Optional |
| `cumulative_type_params::grain_to_date` | Sets the accumulation grain, such as `month`, which will accumulate data for one month and then restart at the beginning of the next. This can't be used with `window`. | Optional |

> | `cumulative_type_params::period_agg` | Specifies how to aggregate the cumulative metric when summarizing data to a different granularity. Options are <br /> - `first` (Takes the first value within the period) <br /> - `last` (Takes the last value within the period <br /> - `avg` (Calculates the average value within the period). <br /> <br /> Defaults to `first` if no `window` is specified. | Optional |

Author: ```suggestion
| `cumulative_type_params::period_agg` | Specifies how to aggregate the cumulative metric when summarizing data to a different granularity. Can be used with grain_to_date. Options are <br /> - `first` (Takes the first value within the period) <br /> - `last` (Takes the last value within the period <br /> - `avg` (Calculates the average value within the period). <br /> <br /> Defaults to `first` if no `window` is specified. | Optional |
```

> | `type_params::measure` | A list of measure inputs. | Required |

Author: ```suggestion
| `type_params::measure` | A dictionary describing the measure you will use. | Required |
```

- `cumulative_order_total_l1m`: Calculates the trailing 1-month cumulative order total. Uses `cumulative_type_params` to specify a `window` of 1 month.

- `cumulative_order_total_mtd`: Calculates the month-to-date cumulative order total, respectively. Uses `cumulative_type_params` to specify a `grain_to_date` of `month`.

<File name='models/marts/sem_semantic_model_name.yml'>

```yaml

metrics:
  - name: cumulative_order_total
    label: Cumulative order total (All-Time)    
    description: The cumulative value of all orders
    type: cumulative
    type_params:
      measure: 
        name: order_total
  
  - name: cumulative_order_total_l1m
    label: Cumulative order total (L1M)   
    description: Trailing 1-month cumulative order total
    type: cumulative
    type_params:
      measure: 
        name: order_total
      cumulative_type_params:
        window: 1 month
  
  - name: cumulative_order_total_mtd
    label: Cumulative order total (MTD)
    description: The month-to-date value of all orders
    type: cumulative
    type_params:
      measure: 
        name: order_total
      cumulative_type_params:
        grain_to_date: month
```
</File>

> ### Period agg

Author: ```suggestion
### Granularity options
```

:::tip
Note that we use the double colon (::) to indicate whether a parameter is nested within another parameter. So for example, `query_params::metrics` means the `metrics` parameter is nested under `query_params`.
:::

| Parameter | Description | Type |
| --------- | ----------- | ---- |
| `name` | The name of the metric. | Required |
| `description` | The description of the metric. | Optional |
| `type` | The type of the metric (cumulative, derived, ratio, or simple). | Required |
| `label` | Required string that defines the display value in downstream tools. Accepts plain text, spaces, and quotes (such as `orders_total` or `"orders_total"`). | Required |
| `type_params` | The type parameters of the metric. | Required |
| `type_param:window` | The accumulation window, such as 1 month, 7 days, 1 year. This can't be used with `grain_to_date`. | Optional  |
| `type_param:grain_to_date` | Sets the accumulation grain, such as `month`, which will accumulate data for one month. Then restart at the beginning of the next. This can't be used with `window`. | Optional |

> | `type_param:period_agg` | Specifies how to roll up the cumulative metric to another granularity.  Options are `start`, `end`, `average`. Defaults to `start` if no `window` is specified. | Optional |

Author: @Jstein77 I'm not sure I udnerstand the definition? And what does teh start/end/avg mean?

> | `type_param:period_agg` | Specifies how to roll up the cumulative metric to another granularity.  Options are `start`, `end`, `average`. Defaults to `start` if no `window` is specified. | Optional |

Author: ```suggestion
| `type_param:period_agg` | Specifies how to roll up the cumulative metric to another granularity.  Options are `first`, `last`, `avg`. Defaults to `start` if no `window` is specified. | Optional |
```

> | `type_param:window_choice` | Specifies whether to take the cumulative metric value from the beginning (min) or end (max) of the window. Defaults to `min`. | Optional |

Author: @Jstein77 is this right? and can window_choice and period_agg be used with window and grain_to_date?

    type: cumulative # Required
    label: The value that will be displayed in downstream tools # Required
    type_params: # Required

>       period_agg: start # Optional. Defaults to start. Accepted values: start|end|average

Author: added exampl here

```yaml

metrics:
  - name: cumulative_order_total
    label: Cumulative order total (All-Time)    
    description: The cumulative value of all orders
    type: cumulative
    type_params:
      measure: 
        name: order_total
  
  - name: cumulative_order_total_l1m
    label: Cumulative order total (L1M)   
    description: Trailing 1-month cumulative order total
    type: cumulative
    type_params:
      measure: 
        name: order_total
      cumulative_type_params:
        window: 1 month
  
  - name: cumulative_order_total_mtd
    label: Cumulative order total (MTD)
    description: The month-to-date value of all orders
    type: cumulative
    type_params:
      measure: 
        name: order_total
      cumulative_type_params:
        grain_to_date: month
```
</File>

### Granularity options
Granularity options for cumulative metrics are slightly different than granularity for other metric types. 
- For other metrics, we use the `date_trunc` function to implement granularity. 
- However, cumulative metrics are non-additive (values can't be added up), so we can't use the `date_trunc` function to change their time grain granularity.

> - Instead, we use the `period_agg` parameter with `first()`, `last()`, and `avg()` functions to aggregate cumulative metrics over the requested period. 

Author: ```suggestion
- Instead, we use the `period_agg` parameter with `first()`, `last()`, and `avg()` SQL functions to aggregate cumulative metrics over the requested period. 
```

- By default, we take the first value of the period. You can change this by specifying a different function using the `period_agg` parameter.

In the following example, we define a cumulative metric, `cumulative_revenue`, that calculates the cumulative revenue for all orders:

<File name='models/marts/sem_semantic_model_name.yml'>

```yaml
- name: cumulative_revenue
  description: The cumulative revenue for all orders.
  label: Cumulative revenue (All Time)
  type: cumulative
  type_params:
    measure: revenue
    cumulative_type_params:
      period_agg: first # Optional. Defaults to first. Accepted values: first|end|avg
```
</File>

In this example, `period_agg` is set to `first`, which chooses the first value for the selected granularity window. To query `cumulative_revenue` by week, use the following query syntax: 
- `dbt sl query --metrics cumulative_revenue --group-by metric_time__week`

<Expandable alt_header="Expand toggle to view how the SQL compiles">

> Note the use of the `window` function to select the `first` value. For `last` and `avg`, we would replace the `first_value()` function with `last_value()` and `average` respectively.

Author: ```suggestion
Note the use of the `window` function to select the `first` value. For `last` and `avg`, we would replace the `first_value()` function in the generated SQL with `last_value()` and `average` respectively.
```

Instead we the first_value(), last_value(), and avg() aggregation functions to aggregate cumulative metrics over the requested period. By default we take the first value of the period. You can change this behaviour using the `period_agg` parameter. 

Let's walk through an example using the following configs:

```yaml
- name: cumulative_revenue
  description: The cumulative revenue for all orders.
  label: Cumulative Revenue (All Time)
  type: cumulative
  type_params:
    measure: revenue
    cumulative_type_params:
      period_agg: first # Optional. Defaults to first. Accepted values: first|end|avg
```

`period_agg` is set to first so we will choose the first value for the select granularity window. Let's say we cumulative_revenue by week using the following query: `dbt sl query --metrics cumulative_revenue --group-by metric_time__week`.

This compiles the following SQl, Note the use of the window function to select the first value:

> 

Author: @Jstein77 i'm not sure i understand the diff btw `first_value()`, etc. here compared to `first`. but is this placed correctly

Reviewer: It's because we use the `first_value()` function in the generated SQL. We can remove this if it's confusing.

      grain_to_date: month
```

## Granularity options
Ensure you can request cumulative metrics at different granularities, reducing redundancy and configuration efforts:

> - Window options

Author: @Jstein77 am i right in laying out the diff types of granularity optios for users? and are these the right ones?

```

We can write a cumulative metric `weekly_customers` as such:

``` yaml
metrics: 
  - name: weekly_customers # Define the measure and the window.
  type: cumulative
  type_params:
    measure: customers
    window: 7 days # Setting the window to 7 days since we want to track weekly active
    window_choice: min # Value at the beginning of the window

>     period_agg: first # ADD CONTEXT

Author: @Jstein77 what would period_agg do here?

```

From the sample YAML above, note the following:

* `type`: Specify cumulative to indicate the type of metric. 

> * `type_params`: Specify the measure you want to aggregate as a cumulative metric. You have the option of specifying a `window`, `grain to date`, `period_agg`, `window_choice`.  

Author: added period_agg and window_choice here

        agg: sum
```

We can compare the difference between a 1-month window and a monthly grain to date. 
- The cumulative metric in a window approach applies a sliding window of 1 month
- The grain to date by month resets at the beginning of each month.

```yaml
metrics:
  - name: cumulative_order_total_l1m  # For this metric, we use a window of 1 month 
    label: Cumulative Order total (L1M)
    description: Trailing 1-month cumulative order amount
    type: cumulative
    type_params:
      measure: order_total
      window: 1 month # Applies a sliding window of 1 month
  - name: cumulative_order_total_mtd   # For this metric, we use a monthly grain-to-date 
    label: Cumulative Order total (MTD)
    description: The month-to-date value of all orders
    type: cumulative
    type_params:
      measure: order_total
      grain_to_date: month # Resets at the beginning of each month

>       window_choice: min # Value at the beginning of the window

Author: @Jstein77 added window_choice here. What is the context for period_agg? shoudl i remove here?

      period_agg: first # ADD CONTEXT
```

> ### Implementation

Author: the 'first' was from the notion example  - shouldn't it be start,end,avg?

To calculate the cumulative value of the metric over a given window, use a time range join to a timespine table using the primary time dimension as the join key. Use the accumulation window in the join to decide whether to include on a particular day. Refer to the following example cumulative metric SQL code:

> ``` sql

Author: do we need to update this sql code @Jstein77?

Reviewer: No we shouldn't have to, since this isn't showing granularity.

        agg: sum
```

We can compare the difference between a 1-month window and a monthly grain to date. 
- The cumulative metric in a window approach applies a sliding window of 1 month
- The grain to date by month resets at the beginning of each month.

```yaml
metrics:
  - name: cumulative_order_total_l1m  # For this metric, we use a window of 1 month 
    label: Cumulative Order total (L1M)
    description: Trailing 1-month cumulative order amount
    type: cumulative
    type_params:
      measure: order_total
      window: 1 month # Applies a sliding window of 1 month
  - name: cumulative_order_total_mtd   # For this metric, we use a monthly grain-to-date 
    label: Cumulative Order total (MTD)
    description: The month-to-date value of all orders
    type: cumulative
    type_params:
      measure: order_total
      grain_to_date: month # Resets at the beginning of each month
      window_choice: min # Value at the beginning of the window
      period_agg: first # ADD CONTEXT
```

### Window choice

> Window choice specifies whether to take the cumulative metric value from the beginning (min) or end (max) of the window. Defaults to `min`.

Author: need to confirm is this is right?

  type: cumulative
  type_params:
    measure: order_count
    grain_to_date: month
    window_choice: min # or max
```

This compiles the following SQL code:

```sql
with staging as (
    select
        subq_3.date_day as metric_time__day,
        date_trunc('week', subq_3.date_day) as metric_time__week,
        sum(subq_1.order_count) as orders_last_month_to_date
    from dbt_jstein.metricflow_time_spine subq_3
    inner join (
        select
            date_trunc('day', ordered_at) as metric_time__day,
            1 as order_count
        from analytics.dbt_jstein.orders orders_src_10000
    ) subq_1
    on (
        subq_1.metric_time__day <= subq_3.date_day
    ) and (
        subq_1.metric_time__day >= date_trunc('month', subq_3.date_day)
    )
    group by
        subq_3.date_day,
        date_trunc('week', subq_3.date_day)
)

select
    *
from (
    select
        metric_time__week,
        first_value(orders_last_month_to_date) over (partition by date_trunc('week', metric_time__day) order by metric_time__day) as cumulative_revenue
    from
        staging
)
group by
    metric_time__week,
    cumulative_revenue
order by
    metric_time__week
    1
```

> ### Period aggregation

Author: need to confirm is this is right?
